### DISCUSSION TITLE
http header의 cache-control에 대해 설명해 주세요

### DISCUSSION THREAD

**Discussion by JUDONGHYEOK** [woowacourse-study/2022-woowahan-bansanghwe#77] [Network]
## cache-control 헤더란?
> Cache-Control 일반 헤더 필드는 요청과 응답 내의 캐싱 메커니즘을 위한 디렉티브를 정하기 위해 사용됩니다. 캐싱 디렉티브는 단방향성이며, 이는 요청 내에 주어진 디렉티브가 응답 내에 주어진 디렉티브와 동일하다는 것을 뜻하지는 않는다는 것을 의미합니다. 

Cache-contol 헤더는 말그대로 캐시를 컨트롤 하기위해 만들어진 헤더입니다. 그렇다면 Cache-control의 옵션은 무엇이 있을까요?
Cache-control은 요청에 diretive가 존재한다고 해서 응답에 똑같이 존재하거나 복사된다는 의미는 아닙니다. 따라서 RFC 9111에서는 Cache-contorl의 헤더를 크게 요청 지령과 응답지령으로 나뉩니다.
다음은 RFC에 명시된 Cache-Control의 directives입니다.

### Request Directives
1. max-age
    max-age 지시문은 클라이언트가 지정된 시간 이하인 응답을 선호한다는 것을 의미합니다. 따라서 캐시관리자로부터 지정된 시간보다 오래된 데이터를 반환하지 말아달라는 요청을 합니다.

2. max-stale 
    max-stale 지시문은 클라이언트가 신선도 기한을 초과한 응답을 수락할 것을 나타냅니다. max-stale이 지정된 경우에는 클라인트는 신선도 기한을 지정된 시간 이하로 초과한 응답을 수락합니다. 만약에 지정되지 않을 경우에 클라이언트는 모든 응답을 수락합니다. 
    예를 들어 max-stale이 5로 지정된 경우 신선도 기한이 초과했더라도 5초 이하의 유호시간이 지난 데이터를 반환하는 것을 허락한다. 라는 의미의 헤더입니다. 
3. min-fresh
    min-fresh는 신선도 기한이 현재 수명에 지정된 시간을 더한 값보다 작지 않은 응답을 클라이언트가 선호함을 나타냅니다. 즉 클라이언트가 최소한 지정된 시간동안 최신 상태로 유지되는 응답을 원합니다.
4. no-cache
   Request에서의 no-cache는 캐시관리자를 거치지 않고 곧장 서버와의 연결을 뜻합니다. 클라이언트가 원본 서버에서의 유효성 검사없이 요청을 충족하기 위해 사용되지 않는 응답을 선호함을 나타냅니다. 따라서 캐시된 복사본을 보여주기 이전에 재검증을 위한 요청을 원본서버로 보내는 것을 강제합니다.
5. no-store
    no-store는 request와 response에서의 그 의미가 동일한데요. 둘 중 어느 부분에서도 저장을 해서는 안된다는 의미입니다. 이 옵션은 공용 캐시나 private캐시 모두 적용됩니다.
7. no-tarnsform
    일부 중개자들은 메시지 및 내용을 변환할 수 있습니다. 예를 들어서 프록시는 캐시 공간을 절약하거나 느린 링크의 트래픽 양들을 줄이기 위해 이미지 형식을 변환할 수 있습니다. no-transform 옵션은 이러한 변환을 클라이언트에서 방지하도록 요청하는 것을 나타냅니다.
9. only-if-cached
    only-if-cached 지시문은 클라이언트가 캐싱된 응답만을 얻기 원한다는 것을 나타냅니다. 이 지시문을 수신할시 캐시는 다른 조건들과 일치하는 캐싱된 응답 또는 504 Gateway-time-out을 반환해야 합니다.

### Response Directives
1. max-age
    응답에서의 max-age는 max-age가 지정된 시간을 초과할시 응답을 오래된 것으로 간주함을 나타냅니다.
3. must-revalidate
    must-revalidate는 캐시가 사용되기 이전에 서버에서 유효성 검사를 수행해야 하며 기존 리소스의 상태를 확인하고 만료된 리소스는 절대 사용되어서는 안됨을 나타내는 지시문입니다. 모든상황에서 캐시는 이 옵션을 무시해서는 안되며 특히 캐시 연결이 끊긴 경우는 캐시는 오래된 응답을 재사용하기 보다는 오류응답을 생성해야 합니다.
5. must-understand
    must-understand는 응답의 상태코드에 대한 요구사항을 이해하고 준수하는 캐시에 대한 응답의 캐싱을 제한하는 의미를 갖고 있습니다. must-understand를 포함하는 응답은 no-store 또한 포함해야 합니다. must-understand 지시문을 구현한 캐시가 이를 포함하는 응답을 수신할 때는 상태코드의 캐싱 요구사항을 이해하고 구현한 경우 no-store 옵션을 무시해야 합니다.
7. no-cache
    no-cache는 웅답이 유효성 검사를 위해 캐시를 사용하기 전에 필수로 server에 유효성 체크를 해야 함을 의미합니다.
9. no-store
    위의 no-store 옵션과 동일합니다.
10. no-transform
    위의 no-transform 옵션과 동일합니다.
11. private
    private는 공유 캐시가 해당 응답을 저장하지 않아야 함을 뜻합니다. 
13. proxy-revalidate
    응답이 오래되면 공유 캐시가 원본 서버에서 성공적으로 유효성 검사를 할 때 까지 해당 응답을 재사용해서는 안됨을 의미합니다.
15. public
    public은 캐시가 응답이 금지되더라도 캐시가 어느 곳에서든 응답을 저장할 수 있음을 나타냅니다.
16. s-maxage
    s-maxage는 공유 캐시의 경우 이 지시문에 의해 지정된 최대 수명이 max-age 지시문 또는 expires 헤더 필드에 의해 지정된 최대 수명을 s-maxage에서 재정의 함을 의미합니다.

### 관련 자료
https://developer.mozilla.org/ko/docs/Web/HTTP/Headers/Cache-Control
https://www.rfc-editor.org/rfc/rfc9111.html#name-cache-control

### 연관 질문
no-store 로도 충분할 것 같은데 no-cache, must-revalidate 는 왜 같이 추가하는 것인가요?
https://stackoverflow.com/questions/49547/how-do-we-control-web-page-caching-across-all-browsers

**Comment by kamwoo** [comment]
cache-control 헤더란?

**Comment by kamwoo** [comment]
캐싱 메커니즘이란?

**Comment by kamwoo** [comment]
헤더란?

**Comment by kamwoo** [comment]
request directive에 대해서 설명해주세요

**Comment by kamwoo** [comment]
무상태성은 무엇을 의미하나요?

**Comment by kamwoo** [comment]
무상태성이라면 사용자를 어떻게 구분하고, 원하는 정보를 전달해줄 수 있을까요?

**Comment by kamwoo** [comment]
쿠키와 세션의 차이점

**Comment by kamwoo** [comment]
프로젝트를 하면서 인증/인가에 어떤 방식을 사용했나요?

**Comment by kamwoo** [comment]
토큰 인증 방식을 사용한 이유가 있나요?

**Comment by kamwoo** [comment]
토큰 방식을 사용했을 때 단점은 무엇인가요?

**Comment by kamwoo** [comment]
토큰을 사용하면서 발생할 수 있는 보안문제를 해결할 방법이 있을까요?

**Comment by kamwoo** [comment]
refresh token의 단점은 없을까요?

**Comment by kamwoo** [comment]
캐싱을 하는 이유는 무엇인가요?

**Comment by kamwoo** [comment]
Cache-Control을 사용해서 브라우저가 캐싱을 하지 못하게 하는 방법이 무엇이 있을까요?

**Comment by kamwoo** [comment]
max-age를 사용해서 캐싱을 하지 않도록 하는 방법은 무엇이 있을까요?

**Comment by kamwoo** [comment]
max-age의 기본값을 얼마인가요?

**Comment by kamwoo** [comment]
캐시 무효화에 대해서 설명해주세요

**Comment by kamwoo** [comment]
동키콩의 캐싱을 설정하는 기준이 있을까요?

**Comment by kamwoo** [comment]
캐싱 기간을 1년으로 설정하는 이유가 있을까요?

**Comment by kamwoo** [comment]
CDN은 무엇인가요?

**Comment by kamwoo** [comment]
max-age와 s-maxage의 차이점에 대해서 설명해주세요
